File: merge-two-sorted-lists.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 *
 * @typedef {Object} ListNode
 * @property {number} val
 * @property {ListNode | null} next
 */

/**
 * @param {ListNode} l1
 * @param {ListNode} l2
 * @return {ListNode}
 */
var mergeTwoLists = function (l1, l2) {

  if (!l1 && !l2) {
    return null;
  }

  let v1 = l1 ? l1.val : 0;
  let v2 = l2 ? l2.val : 0;

  if (l1 && (v1 <= v2 || !l2)) {
    return {
      val: v1,
      next: mergeTwoLists(l1.next, l2)
    };
  }

  return {
    val: v2,
    next: mergeTwoLists(l1, l2 && l2.next)
  };
};

// const a = {val: 1, next: {val: 2, next: {val: 4}}};
// const b = {val: 1, next: {val: 3, next: {val: 4}}};

// const a = {val: -9, next: {val: 3}};
// const b = {val: 5, next: {val: 7}};

const a = {val: 1};
const b = null;

const res = mergeTwoLists(a, b);

function r (l) {
  let res = [];
  let cur = l;
  while (cur) {
    res.push(cur.val);
    cur = cur.next;
  }
  return res;
}

console.log(r(res));

File: py/merge-two-sorted-lists.py
# 21. Merge Two Sorted Lists
# You are given the heads of two sorted linked lists list1 and list2.
# Merge the two lists in a one sorted list. The list should be made by splicing together the nodes of the first two
# lists.
# Return the head of the merged linked list.
#
# Example 1:
# Input: list1 = [1,2,4], list2 = [1,3,4]
# Output: [1,1,2,3,4,4]
#
# Example 2:
# Input: list1 = [], list2 = []
# Output: []
#
# Example 3:
# Input: list1 = [], list2 = [0]
# Output: [0]

# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


from typing import Optional


class Solution:
    def mergeTwoLists(self, list1: Optional[ListNode], list2: Optional[ListNode]) -> Optional[ListNode]:
        head = ListNode(-1)
        temp = head
        cur1, cur2 = list1, list2
        while cur1 and cur2:
            if cur1.val < cur2.val:
                temp.next = cur1
                cur1 = cur1.next
            else:
                temp.next = cur2
                cur2 = cur2.next

            temp = temp.next

        if cur1: temp.next = cur1
        if cur2: temp.next = cur2

        return head.next


if __name__ == '__main__':
    # res = Solution().moveZeroes([4,2,4,0,0,3,0,5,1,0])
    print(1)
    #
    # res = Solution().moveZeroes([1, 0, 1])
    # print(res)
    #
    # res = Solution().moveZeroes([0,1,0,3,12])
    # print(res)
    #
    # res = Solution().moveZeroes([0, 1, 0, 3, 12, 0])
    # print(res)
    #
    # res = Solution().moveZeroes([0])
    # print(res)
